src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs

```diff
@@ -781,25 +810,454 @@ internal BoundExpression BindCollectionExpressionConstructor(SyntaxNode syntax,
+            }
+        }
+
+        private bool HasParamsCollectionTypeInProgress(NamedTypeSymbol toCheck)
         {
-            var collectionCreation = BindCollectionExpressionConstructor(syntax, targetType, diagnostics);
-            return !collectionCreation.HasErrors;
+            Binder? current = this;
+            while (current?.Flags.Includes(BinderFlags.CollectionExpressionConversionValidation) == true)
+            {
+                if (current.ParamsCollectionTypeInProgress?.OriginalDefinition.Equals(toCheck.OriginalDefinition, TypeCompareKind.AllIgnoreOptions) == true)
+                {
+                    // We are in a cycle.
+                    return true;
+                }
+
+                current = current.Next;
+            }
+
+            return false;
         }
 
-        internal bool HasCollectionExpressionApplicableAddMethod(SyntaxNode syntax, TypeSymbol targetType, TypeSymbol elementType, BindingDiagnosticBag diagnostics)
+        internal bool HasCollectionExpressionApplicableAddMethod(SyntaxNode syntax, TypeSymbol targetType, TypeSymbol elementType, out ImmutableArray<MethodSymbol> addMethods, BindingDiagnosticBag diagnostics)
         {
+            Debug.Assert(!targetType.IsDynamic());
+
+            NamedTypeSymbol? namedType = targetType as NamedTypeSymbol;
+
+            if (namedType is not null && HasParamsCollectionTypeInProgress(namedType))
+            {
+                // We are in a cycle. Optimistically assume we have the right Add to break the cycle
+                addMethods = [];
+                return true;
+            }
+
             var implicitReceiver = new BoundObjectOrCollectionValuePlaceholder(syntax, isNewInstance: true, targetType) { WasCompilerGenerated = true };
             var elementPlaceholder = new BoundValuePlaceholder(syntax, elementType) { WasCompilerGenerated = true };
             var addMethodBinder = WithAdditionalFlags(BinderFlags.CollectionInitializerAddMethod | BinderFlags.CollectionExpressionConversionValidation);
-            var result = BindCollectionInitializerElementAddMethod(
-                syntax,
-                ImmutableArray.Create<BoundExpression>(elementPlaceholder),
-                hasEnumerableInitializerType: true,
+
+            if (namedType is not null)
+            {
+                addMethodBinder = new ParamsCollectionTypeInProgressBinder(namedType, addMethodBinder);
+            }
+
+            return bindCollectionInitializerElementAddMethod(
                 addMethodBinder,
+                syntax,
+                elementPlaceholder,
                 diagnostics,
-                implicitReceiver);
-            return !result.HasErrors;
+                implicitReceiver,
+                out addMethods);
+
+            // This is what BindCollectionInitializerElementAddMethod is doing in terms of reporting diagnostics and detecting a failure
```

Comment:
>BindCollectionInitializerElementAddMethod [](http://example.com/codeflow?start=28&length=41)

This seems like a lot of code that needs to stay in sync with `BindCollectionInitializerElementAddMethod`. Is the reason for duplicating the code to handle cycles and return the add methods, and is there an easy way to use the existing `Binder` methods? For instance, if the reason is to handle cycles, could we use throw an exception at the point `HasParamsCollectionTypeInProgress(namedType)` is determined and have the outer most caller that created the `ParamsCollectionTypeInProgressBinder` catch that case?

Reply:
There are several reasons. Main reasons is to reduce amount of binding and have a better control what is causing a failure. And cycles, of course. For example, in general we are not supposed to make semantic decisions based on ```BoundNode.HasErrors``` flag, but we did. We are trying to not set the flag too liberally, but that guideline is not always followed (especially in a very old code), because we are not supposed to use its value for semantic decisions. Also, I did find examples when some errors that are not supposed to fail conversion classification failed it, etc. I prefer to have little code duplication, but have a better control over what we are doing. I do not expect the methods to change much in the future. In the process, in the new code I was able to drop a lot of original code that is not relevant to our scenarios and cause. Therefore, no need to think about "protecting" those code paths. The chosen approach gives me greater confidence that we are doing the right thing from behavior point of view.